src/main/java/com/github/cameltooling/lsp/internal/diagnostic/DiagnosticService.java

```diff
@@ -0,0 +1,145 @@
+ */
+package com.github.cameltooling.lsp.internal.diagnostic;
+
+import java.io.ByteArrayInputStream;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+
+import org.apache.camel.catalog.CamelCatalog;
+import org.apache.camel.catalog.EndpointValidationResult;
+import org.apache.camel.parser.XmlRouteParser;
+import org.apache.camel.parser.model.CamelEndpointDetails;
+import org.eclipse.lsp4j.Diagnostic;
+import org.eclipse.lsp4j.DiagnosticSeverity;
+import org.eclipse.lsp4j.DidSaveTextDocumentParams;
+import org.eclipse.lsp4j.Position;
+import org.eclipse.lsp4j.PublishDiagnosticsParams;
+import org.eclipse.lsp4j.Range;
+import org.eclipse.lsp4j.TextDocumentItem;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.github.cameltooling.lsp.internal.CamelEndpointDetailsWrapper;
+import com.github.cameltooling.lsp.internal.CamelLanguageServer;
+import com.github.cameltooling.lsp.internal.parser.ParserXMLFileHelper;
+import com.github.cameltooling.model.diagnostic.BooleanErrorMsg;
+import com.github.cameltooling.model.diagnostic.CamelDiagnosticEndpointMessage;
+import com.github.cameltooling.model.diagnostic.EnumErrorMsg;
+import com.github.cameltooling.model.diagnostic.IntegerErrorMsg;
+import com.github.cameltooling.model.diagnostic.NumberErrorMsg;
+import com.github.cameltooling.model.diagnostic.ReferenceErrorMsg;
+import com.github.cameltooling.model.diagnostic.UnknownErrorMsg;
+
+public class DiagnosticService {
+	
+	private static final String APACHE_CAMEL_VALIDATION = "Apache Camel validation";
+	private static final Logger LOGGER = LoggerFactory.getLogger(DiagnosticService.class);
+	
+	private CompletableFuture<CamelCatalog> camelCatalog;
+	private CamelLanguageServer camelLanguageServer;
+
+	public DiagnosticService(CompletableFuture<CamelCatalog> camelCatalog, CamelLanguageServer camelLanguageServer) {
+		this.camelCatalog = camelCatalog;
+		this.camelLanguageServer = camelLanguageServer;
+	}
+
+	public void compute(DidSaveTextDocumentParams params) {
+		Map<CamelEndpointDetailsWrapper, EndpointValidationResult> endpointErrors = computeCamelErrors(params);
+		List<Diagnostic> diagnostics = converToLSPDiagnostics(endpointErrors, camelLanguageServer.getTextDocumentService().getOpenedDocument(params.getTextDocument().getUri()));
+		PublishDiagnosticsParams diagnosticParam = new PublishDiagnosticsParams(params.getTextDocument().getUri(), diagnostics);
+		camelLanguageServer.getClient().publishDiagnostics(diagnosticParam);
+	}
+
+	private Map<CamelEndpointDetailsWrapper, EndpointValidationResult> computeCamelErrors(DidSaveTextDocumentParams params) {
+		Map<CamelEndpointDetailsWrapper, EndpointValidationResult> endpointErrors = new HashMap<>();
+
+		if (params.getTextDocument().getUri().endsWith(".xml")) {
```

Comment:
does that mean we only support validation on xml so far?

Reply:
yes.
I created this epic to gather all other closely related feature that would be nice to support soon: https://github.com/camel-tooling/camel-language-server/issues/111